src/netconf_acm.c

```diff
@@ -0,0 +1,1833 @@
+
+    xpath = "/ietf-netconf-acm:nacm/denied-notifications";
+    SR_OPER_SUBSCR(session, mod_name, xpath, sr_nacm_oper_cb);
+
+    return error_info;
+}
+
+API void
+sr_nacm_destroy(void)
+{
+    struct sr_nacm_group *group;
+    struct sr_nacm_rule_list *rule_list, *tmp;
+    uint32_t i, j;
+
+    for (i = 0; i < nacm.group_count; ++i) {
+        group = &nacm.groups[i];
+        free(group->name);
+        for (j = 0; j < group->user_count; ++j) {
+            free(group->users[j]);
+        }
+        free(group->users);
+    }
+    free(nacm.groups);
+
+    LY_LIST_FOR_SAFE(nacm.rule_lists, tmp, rule_list) {
+        free(rule_list->name);
+        for (i = 0; i < rule_list->group_count; ++i) {
+            free(rule_list->groups[i]);
+        }
+        free(rule_list->groups);
+        sr_nacm_remove_rules(rule_list);
+        free(rule_list);
+    }
+
+    pthread_mutex_destroy(&nacm.lock);
+}
+
+/**
+ * @brief Get passwd entry of a user, specifically its UID and GID.
+ *
+ * @param[in] user User to learn about.
+ * @param[out] uid User UID, if set.
+ * @param[out] gid User GID, if set.
+ * @return err_info, NULL on success.
+ */
+static sr_error_info_t *
+sr_nacm_getpwnam(const char *user, uid_t *uid, gid_t *gid)
+{
+    sr_error_info_t *err_info = NULL;
+    struct passwd pwd, *pwd_p;
+    char *buf = NULL;
+    ssize_t buflen;
+    int ret;
+
+    assert(user);
+
+    buflen = sysconf(_SC_GETPW_R_SIZE_MAX);
+    if (buflen == -1) {
+        buflen = 2048;
+    }
+    buf = malloc(buflen);
+    if (!buf) {
+        SR_ERRINFO_MEM(&err_info);
+        return err_info;
+    }
+    ret = getpwnam_r(user, &pwd, buf, buflen, &pwd_p);
+    if (ret) {
+        sr_errinfo_new(&err_info, SR_ERR_OPERATION_FAILED, "Getting user \"%s\" pwd entry failed (%s).", user, strerror(ret));
+        free(buf);
+        return err_info;
+    } else if (!pwd_p) {
+        sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "User \"%s\" not found.", user);
+        free(buf);
+        return err_info;
+    }
+
+    if (uid) {
+        *uid = pwd.pw_uid;
+    }
+    if (gid) {
+        *gid = pwd.pw_gid;
+    }
+    free(buf);
+    return err_info;
+}
+
+/**
+ * @brief Check NACM acces for the data tree. If this check passes, no other check is necessary.
+ * If not, each node must be checked separately to decide.
+ *
+ * @param[in] root Root schema node of the data subtree.
+ * @param[in] user User, whose access to check.
+ * @return non-zero if access allowed, 0 if more checks are required.
+ */
+static int
+sr_nacm_allowed_tree(const struct lysc_node *root, const char *user)
+{
+    sr_error_info_t *err_info;
+    uid_t user_uid;
+
+    /* 1) NACM is off */
+    if (!nacm.enabled) {
+        return 1;
+    }
+
+    /* 2) recovery session allowed */
+    err_info = sr_nacm_getpwnam(user, &user_uid, NULL);
+    if (err_info) {
+         /* FIXME: What if sr_nacm_getpwnam returned an error? What to do now? */
```

Comment:
This is a valid situation for TLS sessions, for example, when the username is not a system user. Why did you change the functionality compared to *netopeer2*? As I have said before, assume it all works correctly unless you have a test that does not work.

It seems I will have to go through the whole file and compare it to what is now in *netopeer2* before merging, you are breaking things with these changes.

Reply:
I assumed that adding more error checking wouldn't hurt, but I probably should've added that in a separate change. I didn't know it would be a problem. Before, `sr_nacm_getpwnam` would just return an int, but I think it does make sense for it to return a standard `sr_error_info_t`, since there can be, for example, allocation errors:
```c
    if (!buf) {
        EMEM;
        return -1;
    }
```
I understand that in the old code, errors would be skipped, because in some situations, it's fine that `sr_nacm_getpwnam` failed. But it also means that the allocation error is also skipped. So, maybe the function should be changed to somehow distinguish when something is an actual error? In the end, the change might be as simple as changing this code:
```c
    if (ret) {
        sr_errinfo_new(&err_info, SR_ERR_OPERATION_FAILED, "Getting user \"%s\" pwd entry failed (%s).", user, strerror(ret));
        free(buf);
        return err_info;
```
to something like this:
```c
    if (ret) {
        SR_LOG_WRN("Getting user \"%s\" pwd entry failed (%s).", user, strerror(ret));
        free(buf);
        return err_info;
```
And then introduce the old return value as an output argument. I would also invert the meaning, the output parameter would be called "found" (1 would be "found" and 0 would be "not found").

The caller would then look something like this:
```c
err_info = sr_nacm_getpwnam(user, &user_uid, NULL, &found);
if (err_info) {
    return 0;
}

if (found && user_uid == SR_SU_UID) {
    return 1;
}
```

Also: even `sr_nacm_allowed_tree` can have an error to return. So maybe the return value should also be `sr_error_info_t` (and its old return value should changed to an output argument)?

Reply:
Yes, memory allocation errors were ignored before. I suppose it makes sense to check them now (because *sysrepo* is a library) but then I would expect the whole `sr_nacm_allowed_tree()` to return err_info and have an output argument `allowed` or something.

Reply:
Okay, I have changed the return values now and (hopefully) restored the old functionality. I have made a separate commit so that the changes can be seen, but I can squash it into the "NACM: Use sr_error_info_t for returns" commit, if you preferred that.

Reply:
It would probably be best to squash all your commits before merging so do not worry about it now.